src/envs/cover.py

```diff
@@ -435,81 +437,171 @@ def action_space(self) -> Box:
+
+        if (l1[0] >= r2[0] or l2[0] >= r1[0]):  # one rect on left side of other
+            return False
+        if (r1[1] >= l2[1] or r2[1] >= l1[1]):  # one rect above the other
+            return False
+        return True
+
     def simulate(self, state: State, action: Action) -> State:
         # Since the action space is lower level, we need to write
         # a lower level simulate function.
         assert self.action_space.contains(action.arr)
+
+        # Get data needed to collision check the trajectory.
         dx, dy, dgrip = action.arr
         next_state = state.copy()
         x = state.get(self._robot, "x")
         y = state.get(self._robot, "y")
         grip = state.get(self._robot, "grip")
-        # Update robot state based on action.
+        held_block = None
+        for block in self._blocks:
+            if state.get(block, "grasp") != -1:
+                assert held_block is None
+                held_block = block
+                hx, hy = state.get(held_block, "x"), \
+                           state.get(held_block, "y")
+                hw, hh = state.get(held_block, "width"), \
+                         state.get(held_block, "height")
+
+        # Ensure neither the gripper nor the possible held block collide with
+        # another block or go below the y-axis.
+        if y + dy < 0 - self.collision_threshold:
+            return state.copy()
+        if held_block is not None:
+            if hy - hh + dy < 0 - self.collision_threshold:
+                return state.copy()
+        p1 = (x, y)
+        p2 = (x + dx, y + dy)
+        for block in self._blocks:
+            bx, by = state.get(block, "x"), state.get(block, "y")
+            bw, bh = state.get(block, "width"), state.get(block, "height")
+            ct = self.collision_threshold
+            # These segments defines a slightly smaller rectangle to prevent
+            # floating point arithmetic making us declare a false positive.
+            segments = [
+                ((bx-bw/2+ct, by-ct), (bx+bw/2-ct, by-ct)),  # top
+                ((bx+bw/2-ct, by-ct), (bx+bw/2-ct, by-bh)),  # right
+                ((bx-bw/2+ct, by-ct), (bx-bw/2+ct, by-bh))  # left
+            ]
+            # Only do the check for blocks that are not held.
+            if held_block is None or \
+                (held_block is not None and block != held_block):
+                if any(self.intersects(p1, p2, p3, p4) for p3, p4 in segments):
+                    return state.copy()
+            # To check block-block collisions we do: for each of the four
+            # vertices of our held block, construct the line segment from its
+            # old location to its new location, and check if this line segment
+            # intersects any segment of the rectangle. Also check if the
+            # segments at the final point intersect any segment of the
+            # rectangle.
+            if held_block is not None and block != held_block:
```

Comment:
can also remove the `held_block is None` here after the suggested refactor

Reply:
note that under the above refactor, this if statement actually changes to just `if held_block is not None:`